docs/source/more_types.rst

```diff
@@ -933,3 +933,210 @@ and non-required keys, such as ``Movie`` above, will only be compatible with
+   class Snowflake:
+       """An absolutely unique object in the database"""
+       def __init__(self) -> None:
+           self.id = uuid.uuid4()
+
+   # 1000 lines later
+
+   class User(Snowflake):
+       id = uuid.uuid4()  # This has valid type, but the meaning
+                          # may be wrong
+
+Some other use cases might be solved by using ``@property``, but note that both
+above use cases can't be solved this way. For such situations, one might want
+to use ``typing.Final``.
+
+Definition syntax
+-----------------
+
+The ``typing.Final`` type qualifier indicates that a given name or attribute
+should never be re-assigned, re-defined, nor overridden. It can be used in
+one of these forms:
+
+* The simplest one is ``ID: Final = 1``. Note that unlike gor generic classes
+  this is *not* the same as ``Final[Any]``. Here mypy will infer type ``int``.
+
+* An explicit type ``ID: Final[float] = 1`` can be used as in any
+  normal assignment.
+
+* In stub files one can omit the right hand side and just write
+  ``ID: Final[float]``.
+
+* Finally, one can define ``self.id: Final = 1`` (also with a type argument),
+  but this is allowed *only* in ``__init__`` methods.
+
+Definition rules
+----------------
+
+The are two rules that should be always followed when defining a final name:
+
+* There can be *at most one* final declaration per module or class for
+  a given attribute:
+
+  .. code-block:: python
+
+     ID: Final = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         id: Final = 1
+         def __init__(self, x: int) -> None:
+             self.id: Final = x  # Error!
+
+  Note that mypy has a single namespace for a class. So there can't be two
+  class-level and instance-level constants with the same name.
+
+* There must be *exactly one* assignment to a final attribute:
+
+  .. code-block:: python
+
+     ID = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         ID = 1
+         ID: Final = 2  # Error!
+
+.. note::
+   Conditional final declarations and final declarations within loops are
+   not supported.
+
+Using final attributes
+----------------------
+
+As a result of a final declaration mypy strives to provide the
+two following guarantees:
+
+* A final attribute can't be re-assigned (or otherwise re-defined), both
+  internally and externally:
+
+  .. code-block:: python
+
+     # file mod.py
+     from typing import Final
+
+     ID: Final = 1
+
+     # file main.py
+     from typing import Final
+
+     import mod
+     mod.ID = 2  # Error, can't assign to constant.
+
+     class SomeCls:
+         ID: Final = 1
+
+         def meth(self) -> None:
+             self.ID = 2  # Error, can't assign to final attribute
+
+     class DerivedCls(SomeCls):
+         ...
+
+     DerivedCls.ID = 2  # Error!
+     obj: DerivedCls
+     obj.ID = 2  # Error!
+
+* A final attribute can't be overridden by a subclass (even with another
+  explicit final declaration). Note however, that final attributes can
```

Comment:
As discussed above, I'm not sure about this restriction. Overriding a class-level final attribute with another class-level final attribute seems like a reasonable thing to do for a per-class constant. This would also allow safe covariant overriding of class-level attributes. Maybe we could even somehow allow covariantly overriding a final attribute in an ABC as a final instance attribute.

Hypothetical example:

```
class Model:
    TABLE_NAME: Final[str]  # It would be better if this was explicitly abstract

class Person(Model):
    TABLE_NAME: Final = 'Persons'  # Inferred type should be 'str'?

class Address(Model):
    TABLE_NAME: Final = 'Address'
```

Finally, should it be possible to override a (possibly abstract) property with a final attribute?